common/telemetry/config.go

```diff
@@ -0,0 +1,427 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	sdkmetricexp "go.opentelemetry.io/otel/sdk/metric/export"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials/insecure"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	// the following defaults were taken from the grpc docs as of grpc v1.46.
+	// they are not available programatically
+
+	defaultReadBufferSize    = 32 * 1024
+	defaultWriteBufferSize   = 32 * 1024
+	defaultMinConnectTimeout = 10 * time.Second
+
+	// the following defaults were taken from the otel library as of v1.7.
+	// they are not available programatically
+
+	retryDefaultEnabled         = true
+	retryDefaultInitialInterval = 5 * time.Second
+	retryDefaultMaxInterval     = 30 * time.Second
+	retryDefaultMaxElapsedTime  = 1 * time.Minute
+)
+
+type (
+	metadata struct {
+		Name   string
+		Labels map[string]string
+	}
+
+	connection struct {
+		Kind     string
+		Metadata metadata
+		Spec     interface{} `yaml:"-"`
+	}
+
+	grpcconn struct {
+		Endpoint      string
+		Block         bool
+		ConnectParams struct {
+			MinConnectTimeout time.Duration `yaml:"min_connect_timeout"`
+			Backoff           struct {
+				BaseDelay  time.Duration `yaml:"base_delay"`
+				Multiplier float64
+				Jitter     float64
+				MaxDelay   time.Duration `yaml:"max_delay"`
+			}
+		} `yaml:"connect_params"`
+		UserAgent       string `yaml:"user_agent"`
+		ReadBufferSize  int    `yaml:"read_buffer_size"`
+		WriteBufferSize int    `yaml:"write_buffer_size"`
+		Authority       string
+		Insecure        bool
+
+		cc *grpc.ClientConn
+	}
+
+	exporter struct {
+		Kind struct {
+			Signal string
+			Model  string
+			Proto  string
```

Comment:
I don't like `Proto` word here. It means something else in this project :-)

Reply:
Renamed to Protocol/protocol